server/src/services/documentSymbol/onDocumentSymbol.ts

```diff
@@ -0,0 +1,122 @@
+import { RuleNode } from "@nomicfoundation/slang/cst";
+import { ServerState } from "../../types";
+import { SymbolTreeBuilder } from "./SymbolTreeBuilder";
+import { StructDefinition } from "./visitors/StructDefinition";
+import { StructMember } from "./visitors/StructMember";
+import { InterfaceDefinition } from "./visitors/InterfaceDefinition";
+import { FunctionDefinition } from "./visitors/FunctionDefinition";
+import { ContractDefinition } from "./visitors/ContractDefinition";
+import { EventDefinition } from "./visitors/EventDefinition";
+import { StateVariableDeclaration } from "./visitors/StateVariableDeclaration";
+import { VariableDeclaration } from "./visitors/VariableDeclaration";
+import { ConstantDefinition } from "./visitors/ConstantDefinition";
+import { ConstructorDefinition } from "./visitors/ConstructorDefinition";
+import { EnumDefinition } from "./visitors/EnumDefinition";
+import { ErrorDefinition } from "./visitors/ErrorDefinition";
+import { FallbackFunctionDefinition } from "./visitors/FallbackFunctionDefinition";
+import { LibraryDefinition } from "./visitors/LibraryDefinition";
+import { ModifierDefinition } from "./visitors/ModifierDefinition";
+import { ReceiveFunctionDefinition } from "./visitors/ReceiveFunctionDefinition";
+import { UserDefinedValueTypeDefinition } from "./visitors/UserDefinedValueTypeDefinition";
+import { SymbolVisitor } from "./SymbolVisitor";
+
+export function onDocumentSymbol(serverState: ServerState) {
+  return async (
+    params: DocumentSymbolParams
+  ): Promise<DocumentSymbol[] | SymbolInformation[] | null> => {
+    const { telemetry, solcVersions } = serverState;
+    return telemetry.trackTimingSync("onDocumentSymbol", (transaction) => {
+      const { uri } = params.textDocument;
+
+      // Find the file in the documents collection
+      const document = serverState.documents.get(uri);
+
+      if (document === undefined) {
+        throw new Error(`${uri} not found in documents`);
+      }
+
+      const text = document.getText();
+
+      // Get the document's solidity version
+      let span = transaction.startChild({ op: "solidity-analyzer" });
+      const { versionPragmas } = analyze(text);
+      span.finish();
+
+      versionPragmas.push("<= 0.8.19"); // latest supported by slang
+
+      const solcVersion =
+        semver.maxSatisfying(solcVersions, versionPragmas.join(" ")) ||
+        _.last(solcVersions);
+
+      // Parse using slang
+      span = transaction.startChild({ op: "slang-parsing" });
+      const language = new Language(solcVersion!);
+
+      const parseOutput = language.parse(
+        ProductionKind.SourceUnit,
+        document.getText()
+      );
+
+      const parseTree = parseOutput.parseTree;
+      span.finish();
+
+      if (parseTree === null) {
+        const strings = parseOutput.errors.map((e: any) =>
+          e.toErrorReport(uri, text, false)
+        );
+
+        throw new Error(`Slang parsing error: ${strings.join("")}`);
```

Comment:
nit: since they are multiline strings each:

```suggestion
        throw new Error(`Slang parsing error:\n${strings.join("\n")}`);
```